src/v/kafka/protocol/batch_reader.cc

```diff
@@ -134,4 +135,35 @@ kafka_batch_adapter batch_reader::consume_batch() {
     return kba;
 }
 
+ss::future<batch_reader::storage_t>
+batch_reader::do_load_slice(model::timeout_clock::time_point tp) {
+    using data_t = model::record_batch_reader::data_t;
+    return ss::do_with(data_t{}, [this, tp](data_t& batches) {
+        const auto resources_exceeded = [this, tp] {
+            return is_end_of_stream() || model::timeout_clock::now() > tp;
+        };
+        const auto consume_one = [this, &batches]() {
+            auto kba = consume_batch();
+            if (likely(kba.v2_format && kba.valid_crc && kba.batch)) {
+                batches.push_back(std::move(*kba.batch));
+                return ss::now();
+            } else {
+                _do_load_slice_failed = true;
+                batches.clear();
+                return ss::make_exception_future<>(exception(
+                  kafka::error_code::corrupt_message,
+                  fmt_with_ctx(
+                    fmt::format,
+                    "Invalid kafka record parsing: {}",
+                    !kba.v2_format   ? "not v2_format"
+                    : !kba.valid_crc ? "invalid crc"
+                                     : "empty batch")));
```

Comment:
this looks great

Reply:
I think I got the correct order - it makes sense that magic is checked before crc, right?

Reply:
does the order affect anything other than what message gets printed? seems like you could also do like `"crc {} magic {} empty {}", crc, magic, empty` or something along those lines?

Reply:
It's just the printed message, but I like your version more.

I considered adding an enum (or flags) for failure mode, and an exception type to hold it - which I could check it in the test. But I thought the tests would be the only place it would be used.

Reply:
ok, so `valid_crc` isn't set if `magic != 2`
```
runtime error: load of value 27, which is not a valid value for type 'const bool'
```
I'll leave it as it was.